src/use-event-callback.ts

```diff
@@ -1,34 +1,80 @@
-import { useEffect, useState, SyntheticEvent } from 'react'
-import { Observable, Subject, noop } from 'rxjs'
+import { useEffect, useMemo, useState, SyntheticEvent } from 'react'
+import { Observable, BehaviorSubject, Subject, noop } from 'rxjs'
 
-export type EventCallbackState<T, U> = [((e: SyntheticEvent<T>) => void) | typeof noop, U]
+export type VoidAsNull<T> = T extends void ? null : T
 
-export type EventCallback<T, U> = (eventSource$: Observable<SyntheticEvent<T>>) => Observable<U>
+export type EventCallbackState<_T, E, U, I = void> = [
+  (e: E) => void,
+  [U extends void ? null : U, BehaviorSubject<U | null>, BehaviorSubject<I | null>]
+]
+export type ReturnedState<T, E, U, I> = [EventCallbackState<T, E, U, I>[0], EventCallbackState<T, E, U, I>[1][0]]
 
-export function useEventCallback<T, U = void>(callback: EventCallback<T, U>): EventCallbackState<T, U | null>
-export function useEventCallback<T, U = void>(callback: EventCallback<T, U>, initialState: U): EventCallbackState<T, U>
+export type EventCallback<_T, E, U, I> = I extends void
+  ? (eventSource$: Observable<E>, state$: Observable<U>) => Observable<U>
+  : (eventSource$: Observable<E>, inputs$: Observable<I>, state$: Observable<U>) => Observable<U>
 
-export function useEventCallback<T, U = void>(
-  callback: EventCallback<T, U>,
+export function useEventCallback<T, E extends SyntheticEvent<T>, U = void>(
+  callback: EventCallback<T, E, U, void>,
+): ReturnedState<T, E, U | null, void>
+export function useEventCallback<T, E extends SyntheticEvent<T>, U = void>(
+  callback: EventCallback<T, E, U, void>,
+  initialState: U,
+): ReturnedState<T, E, U, void>
+export function useEventCallback<T, E extends SyntheticEvent<T>, U = void, I = void>(
+  callback: EventCallback<T, E, U, I>,
+  initialState: U,
+  inputs: I,
+): ReturnedState<T, E, U, I>
+
+export function useEventCallback<T, E extends SyntheticEvent<T>, U = void, I = void>(
+  callback: EventCallback<T, E, U, I>,
   initialState?: U,
-): EventCallbackState<T, U | null> {
-  const initialValue = typeof initialState !== 'undefined' ? initialState : null
-  const [state, setState] = useState<EventCallbackState<T, U | null>>([noop, initialValue])
+  inputs?: I,
+): ReturnedState<T, E, U | null, I> {
```

Comment:
Problems are quite same to `useObservable`.
1. maybe more atomic `useState`?
2. only use hooks in top levels.